CodeGeneration/Sources/SyntaxSupport/CompatibilityLayers.swift

```diff
@@ -0,0 +1,192 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// Computes and caches information about properties and initializers that ought to be generated for the compatibility layer.
+public struct CompatibilityLayers /*: Sendable*/ {
+  /// Properties that are needed in the compatibility layer, in the order they ought to appear in the generated file.
+  public var deprecatedVarsByNode: [SyntaxNodeKind: [Child]] = [:]
+
+  /// Initializer signatures that are needed in the compatibility layer, in the order they ought to appear in the generated file.
+  public var deprecatedInitSignaturesByNode: [SyntaxNodeKind: [InitSignature]] = [:]
+
+  /// Properties that are needed in the compatibility layer, in the order they ought to appear in the generated file.
+  public var deprecatedVarsByTrait: [String: [Child]] = [:]
+
+  /// Initializer signatures that are needed in the compatibility layer, in the order they ought to appear in the generated file.
+  public var deprecatedInitSignaturesByTrait: [String: [InitSignature]] = [:]
+
+  internal init(nodes: [Node], traits: [Trait]) {
+    for node in nodes {
+      realizeLayers(for: node)
+    }
+    for trait in traits {
+      realizeLayers(for: trait)
+    }
+  }
+
+  private func extractedChildren(in layoutNode: LayoutNode) -> ArraySlice<Child> {
+    var children = layoutNode.children[...]
+    if children.first?.isUnexpectedNodes ?? false {
+      children.removeFirst()
+    }
+    if children.last?.isUnexpectedNodes ?? false {
+      children.removeLast()
+    }
+    return children
+  }
```

Comment:
I would inline this. This function only makes sense with the comment on the only call side for it, so I think it’s easier to understand if the implementation lives there as well.